### ISSUE TITLE
Jenkins recieves lots of untagged (#null) webhooks from merged but not closed MRs

### ISSUE THREAD

**Issue by eskobarba** [jenkinsci/gitlab-plugin#309]
Last night it received 16 hooks to build on a MR that was closed 9 months ago.

Today it still receives hooks to build from this old, closed MR.

Configuration in gitlab webhook: Push+Merge
Configuration in jenkins: 
build on merge
rebuild on push to source or target

example, test webhook from gitlab, triggers many MRs that are merged already.  but they are not closed... 

<pre>
INFO: WebHook called with url: /project/project-alpha
May 11, 2016 11:20:16 AM com.dabsquared.gitlabjenkins.trigger.handler.push.OpenMergeRequestPushHookTriggerHandler handleMergeRequest
INFO: project-alpha triggered for push to target branch of open merge request #3,992.
May 11, 2016 11:20:26 AM hudson.model.Run execute
INFO: project-alpha #60 main build action completed: SUCCESS
May 11, 2016 11:20:28 AM com.dabsquared.gitlabjenkins.webhook.GitLabWebHook getDynamic
INFO: WebHook called with url: /project/prisjakt
May 11, 2016 11:20:31 AM com.dabsquared.gitlabjenkins.trigger.handler.push.OpenMergeRequestPushHookTriggerHandler handleMergeRequest
INFO: prisjakt triggered for push to target branch of open merge request #4,001.
May 11, 2016 11:20:31 AM com.dabsquared.gitlabjenkins.trigger.handler.push.OpenMergeRequestPushHookTriggerHandler handleMergeRequest
INFO: prisjakt triggered for push to target branch of open merge request #3,995.
May 11, 2016 11:20:31 AM com.dabsquared.gitlabjenkins.trigger.handler.push.OpenMergeRequestPushHookTriggerHandler handleMergeRequest
INFO: prisjakt triggered for push to target branch of open merge request #3,993.
May 11, 2016 11:20:31 AM com.dabsquared.gitlabjenkins.trigger.handler.push.OpenMergeRequestPushHookTriggerHandler handleMergeRequest
INFO: prisjakt triggered for push to target branch of open merge request #3,991.
May 11, 2016 11:20:32 AM com.dabsquared.gitlabjenkins.trigger.handler.push.OpenMergeRequestPushHookTriggerHandler handleMergeRequest
INFO: prisjakt triggered for push to target branch of open merge request #3,990.
May 11, 2016 11:20:32 AM com.dabsquared.gitlabjenkins.trigger.handler.push.OpenMergeRequestPushHookTriggerHandler handleMergeRequest
INFO: prisjakt triggered for push to target branch of open merge request #3,988.
May 11, 2016 11:20:32 AM com.dabsquared.gitlabjenkins.trigger.handler.push.OpenMergeRequestPushHookTriggerHandler handleMergeRequest
INFO: prisjakt triggered for push to target branch of open merge request #3,987.
May 11, 2016 11:20:32 AM com.dabsquared.gitlabjenkins.trigger.handler.push.OpenMergeRequestPushHookTriggerHandler handleMergeRequest
INFO: prisjakt triggered for push to target branch of open merge request #3,981.
May 11, 2016 11:20:33 AM com.dabsquared.gitlabjenkins.trigger.handler.push.OpenMergeRequestPushHookTriggerHandler handleMergeRequest
INFO: prisjakt triggered for push to target branch of open merge request #3,979.
May 11, 2016 11:20:33 AM com.dabsquared.gitlabjenkins.trigger.handler.push.OpenMergeRequestPushHookTriggerHandler handleMergeRequest
INFO: prisjakt triggered for push to target branch of open merge request #3,931.
May 11, 2016 11:20:33 AM com.dabsquared.gitlabjenkins.trigger.handler.push.OpenMergeRequestPushHookTriggerHandler handleMergeRequest
INFO: prisjakt triggered for push to target branch of open merge request #3,884.
May 11, 2016 11:20:33 AM com.dabsquared.gitlabjenkins.trigger.handler.push.OpenMergeRequestPushHookTriggerHandler handleMergeRequest
INFO: prisjakt triggered for push to target branch of open merge request #3,882.
May 11, 2016 11:20:33 AM com.dabsquared.gitlabjenkins.trigger.handler.push.OpenMergeRequestPushHookTriggerHandler handleMergeRequest
INFO: prisjakt triggered for push to target branch of open merge request #3,859.
May 11, 2016 11:20:33 AM com.dabsquared.gitlabjenkins.trigger.handler.push.OpenMergeRequestPushHookTriggerHandler handleMergeRequest
INFO: prisjakt triggered for push to target branch of open merge request #3,834.
May 11, 2016 11:20:33 AM com.dabsquared.gitlabjenkins.trigger.handler.push.OpenMergeRequestPushHookTriggerHandler handleMergeRequest
INFO: prisjakt triggered for push to target branch of open merge request #3,833.
May 11, 2016 11:20:33 AM com.dabsquared.gitlabjenkins.trigger.handler.push.OpenMergeRequestPushHookTriggerHandler handleMergeRequest
INFO: prisjakt triggered for push to target branch of open merge request #3,825.
May 11, 2016 11:20:33 AM com.dabsquared.gitlabjenkins.trigger.handler.push.OpenMergeRequestPushHookTriggerHandler handleMergeRequest
INFO: prisjakt triggered for push to target branch of open merge request #3,802.
May 11, 2016 11:20:34 AM com.dabsquared.gitlabjenkins.trigger.handler.push.OpenMergeRequestPushHookTriggerHandler handleMergeRequest
INFO: prisjakt triggered for push to target branch of open merge request #3,799.
May 11, 2016 11:20:34 AM com.dabsquared.gitlabjenkins.trigger.handler.push.OpenMergeRequestPushHookTriggerHandler handleMergeRequest
INFO: prisjakt triggered for push to target branch of open merge request #3,760.
May 11, 2016 11:20:34 AM com.dabsquared.gitlabjenkins.trigger.handler.push.OpenMergeRequestPushHookTriggerHandler handleMergeRequest
INFO: prisjakt triggered for push to target branch of open merge request #3,712.
May 11, 2016 11:20:34 AM com.dabsquared.gitlabjenkins.trigger.handler.push.OpenMergeRequestPushHookTriggerHandler handleMergeRequest
INFO: prisjakt triggered for push to target branch of open merge request #3,672.
May 11, 2016 11:20:34 AM com.dabsquared.gitlabjenkins.trigger.handler.push.OpenMergeRequestPushHookTriggerHandler handleMergeRequest
INFO: prisjakt triggered for push to target branch of open merge request #3,585.
May 11, 2016 11:20:34 AM com.dabsquared.gitlabjenkins.trigger.handler.push.OpenMergeRequestPushHookTriggerHandler handleMergeRequest
INFO: prisjakt triggered for push to target branch of open merge request #3,401.
May 11, 2016 11:20:34 AM com.dabsquared.gitlabjenkins.trigger.handler.push.OpenMergeRequestPushHookTriggerHandler handleMergeRequest
INFO: prisjakt triggered for push to target branch of open merge request #3,235.
May 11, 2016 11:20:34 AM com.dabsquared.gitlabjenkins.trigger.handler.push.OpenMergeRequestPushHookTriggerHandler handleMergeRequest
INFO: prisjakt triggered for push to target branch of open merge request #3,172.
May 11, 2016 11:20:34 AM com.dabsquared.gitlabjenkins.trigger.handler.push.OpenMergeRequestPushHookTriggerHandler handleMergeRequest
INFO: prisjakt triggered for push to target branch of open merge request #2,939.
May 11, 2016 11:20:35 AM com.dabsquared.gitlabjenkins.trigger.handler.push.OpenMergeRequestPushHookTriggerHandler handleMergeRequest
INFO: prisjakt triggered for push to target branch of open merge request #2,896.
May 11, 2016 11:20:35 AM com.dabsquared.gitlabjenkins.trigger.handler.push.OpenMergeRequestPushHookTriggerHandler handleMergeRequest
INFO: prisjakt triggered for push to target branch of open merge request #2,884.
May 11, 2016 11:20:35 AM com.dabsquared.gitlabjenkins.trigger.handler.push.OpenMergeRequestPushHookTriggerHandler handleMergeRequest
INFO: prisjakt triggered for push to target branch of open merge request #2,824.
May 11, 2016 11:20:35 AM com.dabsquared.gitlabjenkins.trigger.handler.push.OpenMergeRequestPushHookTriggerHandler handleMergeRequest
INFO: prisjakt triggered for push to target branch of open merge request #2,743.
May 11, 2016 11:20:35 AM com.dabsquared.gitlabjenkins.trigger.handler.push.OpenMergeRequestPushHookTriggerHandler handleMergeRequest
INFO: prisjakt triggered for push to target branch of open merge request #2,617.
May 11, 2016 11:20:35 AM com.dabsquared.gitlabjenkins.trigger.handler.push.OpenMergeRequestPushHookTriggerHandler handleMergeRequest
INFO: prisjakt triggered for push to target branch of open merge request #2,444.
May 11, 2016 11:20:35 AM com.dabsquared.gitlabjenkins.trigger.handler.push.OpenMergeRequestPushHookTriggerHandler handleMergeRequest
INFO: prisjakt triggered for push to target branch of open merge request #2,396.
May 11, 2016 11:20:35 AM com.dabsquared.gitlabjenkins.trigger.handler.push.OpenMergeRequestPushHookTriggerHandler handleMergeRequest
INFO: prisjakt triggered for push to target branch of open merge request #2,306.
May 11, 2016 11:20:35 AM com.dabsquared.gitlabjenkins.trigger.handler.push.OpenMergeRequestPushHookTriggerHandler handleMergeRequest
INFO: prisjakt triggered for push to target branch of open merge request #2,081.
May 11, 2016 11:20:35 AM com.dabsquared.gitlabjenkins.trigger.handler.push.OpenMergeRequestPushHookTriggerHandler handleMergeRequest
INFO: prisjakt triggered for push to target branch of open merge request #468.
</pre>
#468 is merged 2 years ago..

We dont close merge requests,, they just get state merged. gitlab-plugin seems to look for state.opened
eg:
OpenMergeRequestPushHookTriggerHandler.java

```
private List<MergeRequest> getOpenMergeRequests(GitLabApi client, String projectId) {
    List<MergeRequest> result = new ArrayList<>();
    Integer page = 0;
    do {
        Response response = null;
        try {
            response = client.getMergeRequests(projectId, State.opened, page, 100);
```

I see that state.reopened and state.updated gets added when i use the rebuild option.. but none of these states returns MRs with state=merged (^).
How come state.merged gets sent as hooks? when client.getMergeRequests has state.opened?

gitlab: 8.7
jenkins: 1.651.1
gitlab-plugin: 1.2.3-SNAPSHOT

**Comment by coder-hugo** [comment]
Does this still happen for a push to the target branch. If yes, could you please add a logger for "com.dabsquared.gitlabjenkins" with level FINEST (see [README](https://github.com/jenkinsci/gitlab-plugin#user-support)) and paste its output here.

**Comment by eskobarba** [comment]
Its behaving properly since this incident, but i dont think anyone pushed straight to target. My suspicion is that 'test webhook' in gitlab sends a query to the api without a state parameter and then makes all this happen.

I will try to push to target at work tomorrow though.

**Comment by WonderCsabo** [comment]
This should be fixed by https://github.com/jenkinsci/gitlab-plugin/pull/312.

**Comment by omehegan** [comment]
Version 1.2.3 of the plugin has the fix for #312.
